src/core/Akka/Actor/ReceiveActorHandlers.cs

```diff
@@ -0,0 +1,198 @@
+        HandleAny = null;
+    }
+
+    private List<ITypeHandler> TypedHandlers { get; }
+
+    private Action<object>? HandleAny { get; set; }
+
+    private void CanAddMoreHandlers()
+    {
+        if (_hadObjectHandlerWithNoPredicate)
+        {
+            throw new InvalidOperationException("A handler for object with no predicate has already been added. No more handlers can be added as they would be ignored.");
+        }
+
+        if (HandleAny != null)
+        {
+            throw new InvalidOperationException("A handler that catches all messages has been added. No more handlers can be added as they would be ignored.");
+        }
+    }
+    
+    private static ITypeHandler CreateTypeHandler<T>(Predicate<T>? shouldHandlePredicate, Func<T, bool> handler)
+    {
+        if (shouldHandlePredicate == null)
+        {
+            return new TypeHandler<T>(handler);
+        }
+
+        return new PredicateHandler<T>(shouldHandlePredicate, handler);
+    }
+    
+    private static ITypeHandler CreateTypeHandler(Type t, Predicate<object>? shouldHandlePredicate, Func<object, bool> handler)
+    {
+        if (shouldHandlePredicate == null)
+        {
+            return new WeaklyTypedHandler(t, handler);
+        }
+
+        return new WeaklyTypedPredicateHandler(t, shouldHandlePredicate, handler);
+    }
+    
+    public void AddGenericReceiveHandler<T>(Predicate<T>? shouldHandlePredicate, Func<T, bool> handler)
+    {
+        CanAddMoreHandlers();
+        
+        TypedHandlers.Add(CreateTypeHandler(shouldHandlePredicate, handler));
+    }
+    
+
+    public void AddTypedReceiveHandler(Type messageType, Predicate<object>? shouldHandlePredicate, Func<object, bool> handler)
+    {
+        CanAddMoreHandlers();
+        
+        TypedHandlers.Add(CreateTypeHandler(messageType, shouldHandlePredicate, handler));
+
+        // If the message type is object, then we need to track that we have added a handler with no predicate.
+        if (messageType == typeof(object) && 
+            shouldHandlePredicate == null)
+        {
+            _hadObjectHandlerWithNoPredicate = true;
+        }
+    }
+
+    public void AddReceiveAnyHandler(Action<object> handler)
+    {
+        CanAddMoreHandlers();
+
+        HandleAny = handler;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public bool TryHandle(object message)
+    {
+        var messageType = message.GetType();
+        foreach (var handler in TypedHandlers)
+        {
+            if (!handler.TargetType.IsAssignableFrom(messageType)) continue;
+            if (handler.TryHandle(message))
+            {
+                return true;
+            }
+        }
+
+        if (HandleAny == null) return false;
+        HandleAny(message);
+        return true;
+
+    }
+}
+
+internal interface ITypeHandler
+{
+    Type TargetType { get; }
+    
+    bool TryHandle(object message);
+}
+
+internal sealed class WeaklyTypedPredicateHandler : ITypeHandler
```

Comment:
Better to have 4 purpose-built `ITypeHandler`s so we can keep the `TryHandle` as slim as possible. If it turns out that a predicate `null` check has lower perf impact that the interface dispatcher overhead then we can pare this down to 2 implementations. If we get rid of the non-generic `Receive` methods from receive actors then we can get this down to a single implementation and remove the interface altogether.